ttnn/cpp/ttnn/graph/graph_trace_utils.cpp

```diff
@@ -0,0 +1,216 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "graph_trace_utils.hpp"
+
+#include "graph_processor.hpp"
+
+#include "tt_metal/common/assert.hpp"
+
+#include <unordered_set>
+#include <string>
+#include <cstdlib> // std::strtoul
+
+
+namespace ttnn::graph {
+
+namespace {
+ttnn::Shape parse_shape(const std::string_view& shape_string) {
+    // Extract shape values from string like "ttnn.Shape([1, 3, 32, 32])"
+    auto start = shape_string.find('[') + 1;
+    auto end = shape_string.find(']');
+    std::string_view shape_values = shape_string.substr(start, end - start);
+
+    // Vector to hold the parsed shape values
+    std::vector<uint32_t> shape;
+    const char* str = shape_values.data();
+    const char* end_str = str + shape_values.size();
+
+    while (str < end_str) {
+        char* next;
+        uint32_t value = std::strtoul(str, &next, 10);
+        if (str == next) break; // no conversion happened
+        shape.push_back(value);
+        str = next;
+        if (*str == ',') {
+            ++str; // skip the comma
+        }
+        if (*str == ' ') {
+            ++str; // skip spaces, assume a single space
+        }
+    }
+
+    return ttnn::Shape(shape);
+}
+} // namespace
+
+uint32_t extract_peak_memory_usage(const nlohmann::json& trace) {
+    uint32_t total_cb = 0;
+    uint32_t total_buffer = 0;
+    uint32_t peak_memory_usage = 0;
+    std::vector<std::string> current_op;
+
+    for (size_t i = 0; i < trace.size(); ++i) {
+        const auto& v = trace[i];
+
+        if (v["name"] == "function_start") {
+            if (current_op.empty()) {
+                while (++i < trace.size()) {
+                    const auto& inner_v = trace[i];
+                    if (inner_v["name"] == "buffer" && inner_v["params"]["type"] == "L1") {
+                        total_buffer += std::stoi(inner_v["params"]["size"].get<std::string>());
+                    } else if (inner_v["name"].get<std::string>().find("tensor") != std::string::npos) {
+                        continue;
+                    } else {
+                        break;
+                    }
+                }
+                --i;  // adjust for loop increment
+            }
+            current_op.push_back(v["params"]["name"]);
+        } else if (v["name"] == "circular_buffer_allocate") {
+            total_cb += stoi(v["params"]["size"].get<std::string>());
+        } else if (v["name"] == "circular_buffer_deallocate_all") {
+            total_cb = 0;
+        } else if (v["name"] == "buffer_allocate" && v["params"]["type"] == "L1") {
```

Comment:
what if type is DRAM?

Reply:
peak usage counting for L1 only, I will rename method to be clear

Reply:
thats what matters the most at the moment

Reply:
got it, but I'd count for both. Less questions from the code users.

Reply:
Leaving L1 for now